File: LeetCode/remote/lru_cache.cs
//  https://leetcode.com/problems/lru-cache/
//
//   Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and set.
//
//   get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
//   set(key, value) - Set or insert the value if the key is not already present. When the cache reached its capacity,
//   it should invalidate the least recently used item before inserting a new item. 

public class Node {
    public Node prev;
    public Node next;
    public int val;
    public int key;
}

public class LRUCache {
    int Size;
    int Capacity;
    Dictionary<int, Node> cache;
    Node Head;
    Node Tail;
    
    public LRUCache(int capacity) {
        Size = 0;
        Capacity = capacity;
        cache = new Dictionary<int, Node>();
    }

    public int Get(int key) {
        if (cache.ContainsKey(key))
        {
            Head.prev = cache[key];
            if (cache[key].prev != null)
            {
                cache[key].prev = cache[key].next;
            }
            cache[key].next = Head.next;
            Head = cache[key];
            return Head.val;
        }
        
        return -1;
    }

    public void Set(int key, int value) {
        if (cache.ContainsKey(key))
        {
            Head.prev = cache[key];
            if (cache[key].prev != null)
            {
                cache[key].prev = cache[key].next;
            }
            
            cache[key].next = Head.next;
            Head = cache[key];
            cache[key].val = value;
            return;
        }
        
        var node = new Node();
        node.val = value;
        node.key = key;
        node.next = Head;
        if (Head != null)
        {
            Head.prev = node;
        }
        
        if (Tail == null)
        {
            Tail = node;
        }
        
        Head = node;
        cache[key] = Head;
        if (++Size > Capacity)
        {
            Tail.prev.next = null;
            cache.Remove(Tail.key);
            Tail = Tail.prev;
        }
    }
}

File: LeetCode/remote/lru_cache.rb

class LRUCache
    class Node < Struct.new(:key, :val, :prev, :next)
    end

    class Node < Struct.new(:key, :val, :prev, :next)
    end
    
    # Initialize your data structure here
    # @param {Integer} capacity
    def initialize(capacity)
        @head = Node.new(:head)
        @tail = Node.new(:tail)
        @head.next, @tail.prev = @tail, @head
        @hash = {}
        @capacity = capacity
    end

    # @param {Integer} key
    # @return {Integer}
    def get(key)
        return -1 if @hash[key].nil?
        node = @hash[key]
        node.next.prev = node.prev
        node.prev.next = node.next
        node.prev = @head
        node.next = @head.next
        @head.next.prev = node
        @head.next = node
        return node.val
    end

    # @param {Integer} key
    # @param {Integer} value
    # @return {Void}
    def set(key, value)
        if @capacity == 0 && @hash[key].nil?
            target = @tail.prev
            target.prev.next = target.next
            target.next.prev = target.prev
            @hash[target.key] = nil
        else
            @capacity -= 1
        end

        node = @hash[key] || Node.new(key, value)
        node.key, node.val = key, value
        node.next = @head.next
        node.next.prev = node
        node.prev = @head
        @head.next = node
        @hash[key] = node
    end
end

cache = LRUCache.new(1)
cache.set(2, 1)
puts cache.get(2)
cache.print
cache.set(3, 2)
puts cache.get(2)
puts cache.get(3)
